server/src/main/java/com/objectcomputing/checkins/services/action_item/ActionItemServicesImpl.java

```diff
@@ -1,79 +1,121 @@
 import java.util.LinkedHashSet;
 import java.util.Set;
 import java.util.UUID;
 
 import static com.objectcomputing.checkins.util.Util.nullSafeUUIDToString;
 
+@Singleton
 public class ActionItemServicesImpl implements ActionItemServices {
 
-    @Inject
-    private CheckInRepository checkinRepo;
-    @Inject
-    private ActionItemRepository actionItemRepo;
-    @Inject
-    private MemberProfileRepository memberRepo;
+    private final CheckInRepository checkinRepo;
+    private final ActionItemRepository actionItemRepo;
+    private final MemberProfileRepository memberRepo;
+    private final SecurityService securityService;
+    private final CurrentUserServices currentUserServices;
+
+    public ActionItemServicesImpl(CheckInRepository checkinRepo, ActionItemRepository actionItemRepo,
+                                  MemberProfileRepository memberRepo, SecurityService securityService,
+                                  CurrentUserServices currentUserServices) {
+        this.checkinRepo = checkinRepo;
+        this.actionItemRepo = actionItemRepo;
+        this.memberRepo = memberRepo;
+        this.securityService = securityService;
+        this.currentUserServices = currentUserServices;
+    }
 
     public ActionItem save(ActionItem actionItem) {
         ActionItem actionItemRet = null;
+        String workEmail = securityService != null ? securityService.getAuthentication().get().getAttributes().get("email").toString() : null;
+        MemberProfile currentUser = workEmail != null ? currentUserServices.findOrSaveUser(null, workEmail) : null;
+        Boolean isAdmin = securityService != null && securityService.hasRole(RoleType.Constants.ADMIN_ROLE);
+
         if (actionItem != null) {
-            final UUID guildId = actionItem.getCheckinid();
-            final UUID createById = actionItem.getCreatedbyid();
+            final UUID checkinId = actionItem.getCheckinid();
+            final UUID createdById = actionItem.getCreatedbyid();
+
+            CheckIn checkinRecord = checkinRepo.findById(checkinId).orElse(null);
+            Boolean isCompleted = checkinRecord != null ? checkinRecord.isCompleted() : null;
+            final UUID pdlId = checkinRecord != null ? checkinRecord.getPdlId() : null;
+            final UUID teamMemberId = checkinRecord != null ? checkinRecord.getTeamMemberId() : null;
+            validate(checkinId == null || createdById == null, "Invalid action item %s", actionItem);
+            validate(actionItem.getId() != null, "Found unexpected id %s for action item", actionItem.getId());
+            validate(checkinRepo.findById(checkinId).isEmpty(), "CheckIn %s doesn't exist", checkinId);
+            validate(memberRepo.findById(createdById).isEmpty(), "Member %s doesn't exist", createdById);
+            if (!isAdmin && isCompleted) {
+                validate(!currentUser.getId().equals(pdlId) || !currentUser.getId().equals(teamMemberId), "User is unauthorized to do this operation");
```

Comment:
if the user is not admin and checkin is completed, shouldn't we throw ActionItemBadArgException directly?

if the user is not admin and checkin is not completed, then we should validate if user belongs to the associated checkin.

Reply:
This should not be throwing that exception at all. If we are kicking the user out for bad permissions it should be returning a 403. That exception will resolve as a 400.